packages/web-frontend/cypress/scripts/generateConfig/generateReportConfig.js

```diff
@@ -3,34 +3,22 @@
  * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
  */
 
-import fs from 'fs';
 import { uniq } from 'lodash';
 import moment from 'moment';
 
-import { compareAsc, getLoggerInstance, ObjectValidator, stringifyQuery } from '@tupaia/utils';
-import cypressConfig from '../../../cypress.json';
+import { compareAsc, hasContent, isAString, ObjectValidator, stringifyQuery } from '@tupaia/utils';
 import { convertDateRangeToUrlPeriodString } from '../../../src/historyNavigation/utils';
-
-const CONFIG_PATH = 'cypress/config/dashboardReports.json';
-
-const readJsonFile = path => JSON.parse(fs.readFileSync(path, { encoding: 'utf-8' }));
-
-const writeJsonFile = (path, json) => fs.writeFileSync(path, JSON.stringify(json, null, 2));
-
-const isNonEmptyArray = input => Array.isArray(input) && input.length > 0;
-
-const isNonEmptyString = value => {
-  if (typeof value !== 'string' || value.length === 0) {
-    throw new Error('Should be a non empty string');
-  }
+import config from '../../config.json';
+import { readJsonFile } from './helpers';
+
+const objectUrlSchema = {
+  id: [hasContent, isAString],
+  project: [hasContent, isAString],
+  orgUnit: [hasContent, isAString],
+  dashboardGroup: [hasContent, isAString],
 };
```

Comment:
I quite like the declarativeness of our `ObjectValidator`. I personally find schema-based validation having a lot of benefits:

* easy to define accurate schemas
* easier to catch input errors early
* self-documented configuration for devs
* more helpful messages for end users

It seems to me that `ObjectValidator` was mostly used in the earlier days of Tupaia (`meditrak-sever`) but perhaps not many people are aware of tis existence now. That's  a pity because it's so cool!

---

 Side note:

After working in Tamanu [yup](https://www.npmjs.com/package/yup) came in my attention. It is a cute validation library with an elegant syntax. I found it easy to use, the errors seemed a bit more generic than ours but perhaps this is configurable.

I know that we use `ObjectValidator` for some Tupaia-specific checks (ids, async lookups etc). I was wondering whether we could/should leverage an existing library for simpler, `sync` schema validation, like the one in this file? We could do that, or we could generalise `ObjectValidator` a bit - the following cards are about that:

https://github.com/beyondessential/tupaia-backlog/issues/1201
https://github.com/beyondessential/tupaia-backlog/issues/2096

Reply:
Yeah I like `yup`! It could be nice to swap out stuff under the hood of `ObjectValidator` for `yup` and see how that goes, maybe replacing `ObjectValidator` entirely in the end. I definitely agree that schema-based validation is very nice :-)